File: ClauseLibrary.Web/Scripts/Controllers/settingsController.js
// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.  
// See full license at the bottom of this file.

(function (angular) {
    "use strict";
    var app = angular.module('clauseLibraryApp');
    settingsController['$inject'] = [
        '$scope',
        '$filter',
        '$window',
        '$modal',
        'sessionService',
        'settingsService',
        'notificationService',
        'clauseItemService',
        'usSpinnerService',
        'navigationService',
        'dataService',
        'modalService',
        'localDatabaseService'
    ];
    app.controller('settingsController', settingsController);

    function settingsController($scope, $filter, $window, $modal, sessionService, settingsService,
        notificationService, clauseItemService, usSpinnerService, navigationService, dataService, modalService, localDatabaseService) {
        $scope.hasInitialised = false;
        $scope.formSubmitted = false;
        $scope.maximumNameLength = 255;
        $scope.maximumDescriptionLength = 1000;

        $scope.libraries = [];
        $scope.libraryUrl = {
            createLibraryUrl: sessionService.getSpWebUrl().replace(/\/$/, '') + '/clauselibrary'
        };
        $scope.newLibrary = { Name: "Clause Library", Description: "My clauses" };

        $scope.connectableLibraries = [];
        $scope.connectableLibraries.selected = {};
        $scope.connectLibrariesLabel = 'libraries';
        $scope.webUrlConnectedTo = sessionService.getDefaultLibrary();

        $scope.accordion = {
            connect: {
                open: $scope.connectableLibraries.length > 0,
                disabled: $scope.connectableLibraries.length === 0
            },
            create: {
                open: $scope.connectableLibraries.length === 0
            }
        }

        $scope.createLibrarySelectHandler = function(item) {
            $scope.libraryUrl.createLibraryUrl = item.Url;
            console.log('settings library selected', $scope.libraryUrl.createLibraryUrl, item.Url);
        }

        // Used for connecting to libraries
        // Returns an async promise
        $scope.getAllExistingLibraries = function () {

            // start the loading spinner (running by default)
            // usSpinnerService.spin('connect-load');
            return settingsService.getAllExistingLibraries()
                .then(function (data) {

                    $scope.connectableLibraries = data;
                
                    // open the connect panel by default if there are options,
                    // otherwise close and disable
                    var hasConnectableLibraries = data && data.length > 0;
                    $scope.accordion.connect.open = hasConnectableLibraries;
                    $scope.accordion.connect.disabled = !hasConnectableLibraries;
                    $scope.connectLibrariesLabel = data && data.length == 1 ? 'library' : 'libraries';

                    // stop the spinner
                    usSpinnerService.stop('connect-load');

                    console.log('existing', data);

                }).catch(function(error) {
                    console.error(error);

                    notificationService.notify(error, notificationService.types.error);

                    // stop the spinner
                    usSpinnerService.stop('connect-load');
                });
            
        }

        // Used for creating libraries
        // Returns an async promise
        $scope.getPotentialLibraryLocations = function () {
            console.log('requesting libraries');
            
            return settingsService.getPotentialLibraryLocations()
                .then(function (data) {
                    $scope.libraries = data;

                    console.log('getPotentialLibraryLocations', data);

                    // stop the spinner
                    usSpinnerService.stop('create-load');
                })
                        .catch(function(error) {
                    console.error(error);
                    notificationService.notify(error, notificationService.types.error);

                    // stop the spinner
                    usSpinnerService.stop('create-load');
                });
        }

        $scope.create = function (form) {

            $scope.formSubmitted = true;

            // weburl
            var clauseLibrarySourceUrl = $scope.libraryUrl.createLibraryUrl;

            if (form.$invalid) {
                return;
            }

            if ($scope.connectableLibraries.filter(function(library) { return library.HostWebUrl.toLowerCase() === clauseLibrarySourceUrl.toLowerCase() }).length >= 1) {
                notificationService.notify('Library already exists, please enter another URL', notificationService.types.error);
                return;
            }

            // library model
            var library = {
                Name: $scope.newLibrary.Name,
                Description: $scope.newLibrary.Description,
                TenantId: sessionService.getTenantId()
            };

            var notification = notificationService.notify(
                "Creating library '" + library.Name + "' (this may take several moments)...",
                notificationService.types.pending
            );


            settingsService.create(clauseLibrarySourceUrl, library)
                .then(function (data) {
                    var libraryGuid = data;
                    console.log('create library', data);

                    notificationService.update(
                        notification,
                        "Library '" + library.Name + "' successfully created",
                        notificationService.types.success
                    );

                    // Update the settings page with fresh library data
                    $scope.getAllExistingLibraries().then(function() {
                            $scope.connectableLibraries.selected = $filter('filter')($scope.connectableLibraries, { LibraryId: libraryGuid }, true)[0];
                        $scope.connect();
                    });
                })
                .catch(function (error) {
                    console.error(error);
                    notificationService.update(
                        notification,
                        "Failed to create new library '" + library.Name + "'",
                        notificationService.types.error
                    );
                });
        }

        $scope.showControlError = function (formControl) {
            return (formControl.$dirty || $scope.formSubmitted) && formControl.$invalid;
        }

        $scope.cancel = function() {
            navigationService.back();
        }

        $scope.connect = function () {

            var selectedLibraryToConnect = $scope.connectableLibraries.selected;

            if (selectedLibraryToConnect) {
                // library model - only need library id (copied from db)
                var library = {
                    LibraryId: selectedLibraryToConnect.LibraryId,
                    TenantId: selectedLibraryToConnect.TenantId,
                    Name: selectedLibraryToConnect.Name,
                    Description: selectedLibraryToConnect.Description,
                    HostWebUrl: selectedLibraryToConnect.HostWebUrl
                };

                var notification = notificationService.notify(
                    "Connecting to library '" + library.Name + "'...",
                    notificationService.types.pending
                );

                // connects to an existing library for which the current user has sufficient permissions
                // and yeilds a populated library object upon successful connection
                settingsService.connect(library)
                    .then(function(connectedLibrary) {

                        if (connectedLibrary) {
                            // update session service with new web url (will update localStorage)
                            var isUpdated = sessionService.setHostWebUrl(connectedLibrary.HostWebUrl);

                            // update clause item service so that new library data can be displayed in manage view
                            if (isUpdated) {

                                // update local storage
                                sessionService.setNewLibraryConnection(connectedLibrary);

                                // update the visual connected library url in the view
                                $scope.webUrlConnectedTo = connectedLibrary;

                                // we've connected to a new library; 
                                // forcibly refresh the manage view
                                dataService.setSyncRequired(true);

                                notificationService.update(
                                    notification,
                                    "Connected to '" + library.Name + "' successfully",
                                    notificationService.types.success
                                );

                                navigationService.goToManage();
                            } else {
                                notificationService.update(
                                    notification,
                                    "You are already connected to '" + library.Name + "'.",
                                    notificationService.types.info);
                            }
                        } else {
                            notificationService.update(
                                notification,
                                "Something went wrong when trying to connect to " + library.Name,
                                notificationService.types.error
                            );
                        }
                    })
                    .catch(function(error) {
                        console.error(error);
                        notificationService.update(
                            notification,
                            "Failed to connect to library '" + library.Name + "'",
                            notificationService.types.error
                        );
                    });
            } else {
                notificationService.notify("No Library has been selected", notificationService.type.warning);
            }
            
        }

        $scope.clearCache = function() {
        
            var deleteConfirm = $modal.open(modalService.getGenericModal("Are you sure you want to clear the local data from this library?"));

            deleteConfirm.result.then(function(confirmed) {
                if (!confirmed) {
                    return;
                }
                // user confirmed cache clear
                var notification = notificationService.notify("Resetting local store...", notificationService.types.pending);
                localDatabaseService.deleteLocalDB().then(function () {
                    if (sessionService.getHostWebUrl() === $scope.connectableLibraries.selected.HostWebUrl) {
                        dataService.setSyncRequired(true);
                    }

                    notificationService.update(notification, 'Local data store cleared', notificationService.types.success);
                });
            }, function() {
                console.log('Clear cache cancelled');
            });
        }    

        function initialise() {
            $scope.hasInitialised = true;

            $scope.getAllExistingLibraries();
            $scope.getPotentialLibraryLocations();
        }

        initialise();
    };
})(angular);

// ClauseLibrary, https://github.com/OfficeDev/clauselibrary 
//   
// Copyright 2015(c) Microsoft Corporation 
//   
// All rights reserved. 
//   
// MIT License: 
//   
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
// associated documentation files (the "Software"), to deal in the Software without restriction, including 
// without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
// copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the 
// following conditions: 
//   
// The above copyright notice and this permission notice shall be included in all copies or substantial 
// portions of the Software. 
//   
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
// LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT 
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE 
// USE OR OTHER DEALINGS IN THE SOFTWARE. 

File: ClauseLibrary.Web/Controllers/SettingsController.cs
// Copyright (c) Microsoft. All rights reserved. Licensed under the MIT license.  
// See full license at the bottom of this file.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Net.Http;
using ClauseLibrary.Common;
using ClauseLibrary.Web.Models;
using ClauseLibrary.Web.Models.Database.LoginSettings;
using ClauseLibrary.Web.Models.Database.Services;
using ClauseLibrary.Web.Models.DataModel;
using ClauseLibrary.Web.Services;

namespace ClauseLibrary.Web.Controllers
{
    /// <summary>
    /// Provides functionality to control settings used within the application.
    /// </summary>
    public class SettingsController : ApiControllerBase
    {
        private readonly IProvisioningService _provisioningService;
        private readonly ILoginSettingsService _loginSettingsService;

        /// <summary>
        /// Initializes a new instance of the <see cref="SettingsController" /> class.
        /// </summary>
        /// <param name="provisioningService">The provisioning service.</param>
        /// <param name="loginSettingsService">The login settings service.</param>
        public SettingsController(IProvisioningService provisioningService, ILoginSettingsService loginSettingsService)
        {
            _provisioningService = provisioningService;
            _loginSettingsService = loginSettingsService;
        }

        /// <summary>
        /// Gets the subset of the existing libraries for a tenant which the user has access to.
        /// </summary>
        /// <param name="tenantWebUrl">URL of tenant</param>
        /// <param name="userId">string guid of the logged in user. Can be accessed on SharePointAccessInfo</param>
        /// <returns>Enumerable of LibraryModel</returns>
        [System.Web.Http.HttpGet]
        public IEnumerable<LibraryModel> GetAllExistingLibraries(string tenantWebUrl, string userId)
        {
            //get list of libraries from the database; permission checking will be done when
            //the user attempts to connect, not here.
            var libraries = new List<Library>();
            var user = _loginSettingsService.GetUserById(new Guid(userId));
            if (user != null)
            {
                libraries = user.Tenant.Libraries;
            }

            return libraries.Select(library => new LibraryModel(library));
        }

        /// <summary>
        /// Gets a list of locations where the user can create a library (has site create site permissions)
        /// </summary>
        /// <param name="tenantWebUrl">URL of tenant</param>
        /// <param name="accessToken">OAuth Access Token to be used as authentication with SharePoint REST API</param>
        /// <returns>Enumerable of WebModel</returns>
        [System.Web.Http.HttpGet]
        public List<Site> GetPotentialLibraryLocations(string tenantWebUrl, string accessToken = "")
        {
            accessToken = GetAccessToken(accessToken);

            var sites = _provisioningService.GetSitesWhereUserHasPermissions(tenantWebUrl, accessToken,
                SPBasePermissions.FullMask);
            
            return sites;
        }

        /// <summary>
        /// Checks the version of the SharePoint site, and return false if an upgrade is required
        /// </summary>
        /// <param name="webUrl">The web URL.</param>
        /// <param name="accessToken">The access token.</param>
        /// <returns></returns>
        [System.Web.Http.HttpGet]
        public bool SpUpdateRequired(string webUrl, string accessToken = "")
        {
            accessToken = GetAccessToken(accessToken);
            var requiresUpdate = _provisioningService.RequiresUpdate(webUrl, accessToken);

            return requiresUpdate;
        }

        /// <summary>
        /// Upgrades the specified web URL.
        /// </summary>
        /// <param name="webUrl">The web URL.</param>
        /// <param name="accessToken">The access token.</param>
        [System.Web.Http.HttpGet]
        public bool Upgrade(string webUrl, string accessToken = "")
        {
            accessToken = GetAccessToken(accessToken);
            return _provisioningService.Upgrade(webUrl, accessToken);
        }

        /// <summary>
        /// Creates a library (sharepoint site) at specified location.
        /// </summary>
        /// <param name="webUrl">URL where to create the subsite</param>
        /// <param name="accessToken">OAuth Access Token to be used as authentication with SharePoint REST API</param>
        /// <param name="newLibrary">Library to be created</param>
        /// <returns>http Response message</returns>
        [System.Web.Http.HttpPost]
        public HttpResponseMessage Create(string webUrl, Library newLibrary, string accessToken = "")
        {
            accessToken = GetAccessToken(accessToken);
            
            try
            {
                Common.Models.Web libraryWebModel = _provisioningService.CreateClauseLibraryWeb(webUrl, newLibrary.Name, accessToken);

                Library library = new Library
                {
                    LibraryId = new Guid(libraryWebModel.Id),
                    TenantId = newLibrary.TenantId,
                    Name = newLibrary.Name,
                    Description = newLibrary.Description,
                    HostWebUrl = webUrl
                };

                //Create Library in Database
                _loginSettingsService.Add(library);
                _loginSettingsService.Save();

                return new HttpResponseMessage(HttpStatusCode.OK) { Content = new StringContent(library.LibraryId.ToString()) };
            }
            catch (Exception e)
            {
                return new HttpResponseMessage(HttpStatusCode.BadRequest) {ReasonPhrase = e.Message};
            }
        }

        /// <summary>
        /// Connects to existing library (sharepoint site)
        /// </summary>
        /// <param name="tenantWebUrl">webUrl of tenant</param>
        /// <param name="library">The library.</param>
        /// <param name="userId">string guid of the logged in user. Can be accessed on SharePointAccessInfo</param>
        /// <param name="accessToken">OAuth Access Token to be used as authentication with SharePoint REST API</param>
        /// <returns>
        /// http Response message
        /// </returns>
        [System.Web.Http.HttpPut]
        public LibraryConnectionResult Connect(string tenantWebUrl, Library library, string userId, string accessToken = "")
        {
            accessToken = GetAccessToken(accessToken);
            var userEmail = GetCurrentUserEmail();

            //check with provisioning service that they are authoriesed to connect
            var libraryId = library.LibraryId;

            var libraryIdWithReadPermission = _provisioningService.GetWebWhereUserHasPermissions(
                library.HostWebUrl, accessToken, SPBasePermissions.ViewListItems);

            if (libraryIdWithReadPermission == null || libraryIdWithReadPermission.Id == null || new Guid(libraryIdWithReadPermission.Id) != libraryId)
            {
                return null;
            }

            // update db connected User if they have connection
            var user = _loginSettingsService.GetUserById(new Guid(userId));
            if (user != null)
            {
                user.DefaultLibraryId = library.LibraryId;
                _loginSettingsService.Save();

                var accessInfo = new SharePointAccessInfo(library.HostWebUrl)
                {
                    AccessToken = accessToken,
                    UserEmail = userEmail
                };
                accessInfo.Update();

                var libraryConnectionResult = new LibraryConnectionResult
                {
                    Library = library,
                    AccessInfo = accessInfo
                };

                return libraryConnectionResult;
            }
            return null;
        }
    }
}

#region License 
// ClauseLibrary, https://github.com/OfficeDev/clauselibrary 
//   
// Copyright 2015(c) Microsoft Corporation 
//   
// All rights reserved. 
//   
// MIT License: 
//   
// Permission is hereby granted, free of charge, to any person obtaining a copy of this software and 
// associated documentation files (the "Software"), to deal in the Software without restriction, including 
// without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell 
// copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the 
// following conditions: 
//   
// The above copyright notice and this permission notice shall be included in all copies or substantial 
// portions of the Software. 
//   
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT 
// LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT 
// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN 
// ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE 
// USE OR OTHER DEALINGS IN THE SOFTWARE. 
#endregion